File: DeleteXthNodeFromListEnd/DeleteXthNodeFromListEnd.cpp
// DeleteXthNodeFromListEnd.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

void PrintList(ListNode* pHead)
{
    if (!pHead)
        return;
    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }

    std::cout << std::endl;
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

ListNode* GetNodeByIndex(ListNode* pHead, int index)
{
    if (index == -1)
        return nullptr;

    for (int i = 0; i < index; ++i)
    {
        pHead = pHead->next;
    }

    return pHead;
}

int CountLinkedList(ListNode* pHead)
{
    int count = 0;
    while (pHead != nullptr)
    {
        pHead = pHead->next;
        ++count;
    }

    return count;
}

ListNode* RemoveXthNodeFromEnd(ListNode** head, int& x)
{
    if (x == 0)
        return nullptr;

    int count = CountLinkedList(*head);
    if (count < x)
        return nullptr;

    ListNode* pNode = nullptr;
    if (x == count)
    {
        pNode = *head;
        *head = (*head)->next;
        return pNode;
    }

    ListNode* pPrevious = GetNodeByIndex(*head, count - x - 1);
    pNode = pPrevious->next;
    pPrevious->next = pNode->next;
    return pNode;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count == 0)
            break;

        ListNode* pHead = nullptr;
        ListNode* pTail = GetListFromInput(&pHead, count);

        std::cout << "Which node from the end should be deleted? ";
        int xTh = 0;
        std::cin >> xTh;

        ListNode* pDeleted = RemoveXthNodeFromEnd(&pHead, xTh);
        if (pDeleted)
            delete pDeleted;

        std::cout << "After node removed: ";
        PrintList(pHead);
        FreeList(pHead);
    }
}

File: Go/DeleteXthNodeFromListEnd.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) *ListNode {
	if len(v) == 0 {
		return nil
	}

	if len(v) == 1 {
		return &ListNode{v[0], nil}
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead
}

func (pNode *ListNode) Print() {
	fmt.Print(pNode.data)
	if pNode.next != nil {
		fmt.Print(", ")
		pNode.next.Print()
	} else {
		fmt.Println("")
	}
}

func CountList(pNode *ListNode) int {
	count := 0
	for ; pNode != nil; pNode = pNode.next {
		count++
	}

	return count
}

func RemoveXthNodeFromEnd(head *ListNode, x int) *ListNode {
	if head == nil {
		return head
	}

	count := CountList(head)
	if x == count {
		return head.next
	}

	pSaved := head
	for i := 0; i < count-x-1; i++ {
		head = head.next
	}

	head.next = head.next.next
	return pSaved
}

func main() {
	for {
		fmt.Print("Number elements in list: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the nodes in list: ")
		pHead := GetListFromArray(vec)

		fmt.Print("Which node from the end should be removed: ")
		var x int
		fmt.Scan(&x)

		pHead = RemoveXthNodeFromEnd(pHead, x)
		fmt.Print("The resule list: ")
		pHead.Print()
	}
}
